docs/hugo/content/design/ADR-2025-01-ARM-References/_index.md

```diff
@@ -0,0 +1,230 @@
+}
+```
+
+We already have precendent for this, with the `*FromConfig` properties that we inject for selected properties.
+
+Validation would need to ensure the two properties are mutually exclusive.
+
+When constructing the payload to send to ARM, we would need to select between the two properties, using whichever is set. We have an exisiting resource extension point we can use to implement this.
+
+Adding a third option to a boolean flag is a very bad idea (though we've seen it done), so we'd need to rename `$armReference` to something else for clarity. Fortunately, we control `azure_arm.yaml` so this is achievable. (We don't believe any users have their own configuration file, but we should call this out as a breaking in our release notes anyway.)
+
+One possible name for flag would be `$isReference` with values `arm`, `side-by-side` and `no`, though this phrasing is quite awkward.
+
+* PRO: Relatively straightforward
+* PRO: We already have precedent for side-by-side properties
+* PRO: Solves all three scenarios.
+* CON: Problematic to have _two_ properties that handle _three_ use-cases (e.g. use `privateLinkServiceReference` for GVK or ARMId, and `privateLinkServiceId` for alias).
+* CON: No control over the naming of things
+
+### Option 3: Custom reference types
+
+Instead of always replacing the property with a `genruntime.ResourceReference`, allow selection between a set of standard reference types.
+
+For the vast majority of reference properties, we'd continue to use the existing `genruntime.ResourceReference`.
+
+To support the _Alias_ and _Friendly names_ cases, introduce a new reference type, say `genruntime.AliasResourceReference`, with the following structure:
+
+``` go
+type AliasResourceReference struct {
+   Group     string
+   Kind      string
+   Namespace string
+   Name      string
+
+    ARMID string
+
+    Alias string
+}
+```
+
+The code generator would need to be updated to support this new value.
+
+Finally, when the property isn't an reference at all, but just a simple string, we'd have a value that means _leave this alone_. This is the case where we'd currently us `$armReference: false`.
+
+Again, our current `$armReference` flag would need to be renamed to something more general. Perhaps `$referenceType` with possible values `arm`, `armOrAlias`, and `other`.
+
+* PRO: Great YAML structure for end users of resources.
+* PRO: Extensible for the (admittedly unlikely) scenario where we need another different kind.
+* PRO: Cleanly handles both the _Alias_ and _Friendly names_ scenarios.
+* CON: Slightly greater implementation complexity.
+* CON: Does not solve the _Missed ARM References_ problem.
+
+### Option 4: Multivalued custom reference types
+
+Similar to Option 3, but allowing set combinations of values to be specified for the reference type.
+
+Possible values would be `arm`, `alias` and `other`, plus the combinations `arm+alias` and `arm+other`. Other combinations would be invalid.
```

Comment:
As discussed: Consider a yaml list rather than `+`, as it avoids having to support `other+arm` and `arm+other`.

Reply:
Added.